x/twap/logic_test.go

```diff
@@ -521,3 +522,421 @@ func (s *TestSuite) TestPruneRecords() {
+		spOverrides   []spOverride
+		blockTime     time.Time
+
+		expectedMostRecentRecords []expectedResults
+		expectedHistoricalRecords []expectedResults
+		expectError               error
+	}{
+		"no records pre-set; error": {
+			preSetRecords: []types.TwapRecord{},
+			poolId:        1,
+			blockTime:     baseTime,
+
+			expectError: gammtypes.PoolDoesNotExistError{PoolId: 1},
+		},
+		"existing records in different pool; no-op": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId + 1,
+			blockTime:     baseTime.Add(time.Second),
+
+			expectError: gammtypes.PoolDoesNotExistError{PoolId: baseRecord.PoolId + 1},
+		},
+		"the returned number of records does not match expected": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.NewDec(2),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: baseRecord.Asset0Denom,
+					overrideSp: sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: "extra denom B",
+					overrideSp: sdk.NewDecWithPrec(3, 1),
+				},
+			},
+
+			expectError: types.InvalidRecordCountError{Expected: 3, Actual: 1},
+		},
+		"two-asset; pre-set record at t; updated valid spot price": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.NewDec(2),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: baseRecord.Asset0Denom,
+					overrideSp: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+
+			expectedMostRecentRecords: []expectedResults{
+				{
+					spotPriceA: sdk.NewDec(2),
+					spotPriceB: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+			expectedHistoricalRecords: []expectedResults{
+				// The original record.
+				{
+					spotPriceA: baseRecord.P0LastSpotPrice,
+					spotPriceB: baseRecord.P1LastSpotPrice,
+				},
+				// The new record added.
+				{
+					spotPriceA: sdk.NewDec(2),
+					spotPriceB: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+		},
+		"two-asset; pre-set record at t; updated with spot price error in both denom pairs": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:   baseRecord.Asset0Denom,
+					quoteDenom:  baseRecord.Asset1Denom,
+					overrideErr: spError,
+				},
+				{
+					baseDenom:   baseRecord.Asset1Denom,
+					quoteDenom:  baseRecord.Asset0Denom,
+					overrideSp:  sdk.NewDecWithPrec(2, 1),
+					overrideErr: spError,
+				},
```

Comment:
Clarifying question: The description says "updated with spot price error in both denom pairs", there is a spErr in both denoms pairs, but we dont overrideSp for baseQuoteAB. Is this intentional? In my mind, if there is a spot price error in both denom pairs, then they should both return zero (as it currently stands, I am in 100 percent agreement with your recommendation that they should actually return the last known spot price)

Reply:
This is intentional as it currently stands.

We only set the spot price to zero if there is a spot price error and the returned value from amm is an empty `Dec`:
https://github.com/osmosis-labs/osmosis/blob/bcda1cbe01f87111eb9320a8907b6b89d9930fb6/x/twap/logic.go#L49-L51

I'm planning to change this logic in:
https://github.com/osmosis-labs/osmosis/issues/2689